### PR TITLE
Groundwork for loading mzXML files independent of accucor data

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

I intend to implement the loading of mzXML files unassociated with peak annotation data (#905) using the `Peak Annotation Details` sheet in the study doc (as an optional input).  This will allow mzXML files to be associated with database sample records, sequence records, and PeakGroup records.

However, when I was planning this out, I evaluated my plan against my general requirements that a loader class *only* taking in a single data sheet (`Peak Annotation Details`) and a defaults sheet (`Defaults`), and I realized that the `Sequence Number` column did not meet that general requirement, so this PR changes that column to "Sequence Name", composed of a comma-delimited unique value combo representing a single MSRunSequence record.

While I was at it, I also realized that my goal of simplifying the filling-out of the `Sequences` sheet using LC Protocol drop-downs was complicated by the fact that there were multiple LC-protocol-related columns in the sequences sheet.  That meant that either users would have had to make multiple selections that go with one another (to avoid invalid entries and confusing errors) or I would have to write a complex set of excel formulas to link the dropdowns, which I thought seemed uselessly complex, so I decided to split the LC-related data into its own sheet (`LC Protocols`) and use a single `LC Protocol Name` column populated using dropdowns whose values from the `LC Protocols` sheet.

From here, since a discussion about #947 resulted in approval of model changes that would enable #905's implementation to be simpler, I will move on to #947, then on to #905 and #906.

The changes here affect the study doc template.  Here is a new example file (located under [`DataRepo/data/tests/submission_v3/study.xlsx`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/files/15017240/study.xlsx).

## Affected Issues/Pull Requests

- Partially addresses #905
- Partially addresses #906
- Partially addresses #825

## Review Notes

See comments in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/data/tests/submission_v3/defaults.tsv [hunk 1]
```diff
@@ -3,7 +3,5 @@ Sequences	Operator	Xianfeng Zeng
 Peak Annotation Files	Peak Annotation File Type	isocorr
 Peak Annotation Details	Polarity	positive
 Sequences	Date	2020-11-01
-Sequences	LC Protocol	polar-HILIC
-Sequences	LC Run Length	25
-Sequences	LC Description	
+Sequences	LC Protocol Name	polar-HILIC-25-min
```
Comment (hepcat72) [inline DataRepo/data/tests/submission_v3/defaults.tsv:6]: C. The column headers in the Sequences sheet, referred to in the second column of this file (defining default values for those columns), changed.

File: DataRepo/data/tests/submission_v3/sequences.tsv [hunk 2]
```diff
@@ -1,4 +1,4 @@
-Sequence Number	Operator	Date	Instrument	LC Protocol	LC Run Length	LC Description	Notes
-1	Xianfeng Zeng	2021-6-8	HILIC	polar-HILIC	25		
-2	Xianfeng Zeng	2021-10-19	HILIC	polar-HILIC	25		
-3	Xianfeng Zeng	2020-7-22	HILIC	polar-HILIC	25		
+Sequence Name	Operator	Date	Instrument	LC Protocol Name	Notes
+Xianfeng Zeng, polar-HILIC-25-min, QE2, 2021-6-8	Xianfeng Zeng	2021-6-8	QE2	polar-HILIC-25-min	
+Xianfeng Zeng, polar-HILIC-25-min, QE2, 2021-10-19	Xianfeng Zeng	2021-10-19	QE2	polar-HILIC-25-min	
+Xianfeng Zeng, polar-HILIC-25-min, QE2, 2020-7-22	Xianfeng Zeng	2020-7-22	QE2	polar-HILIC-25-min
```
Comment (hepcat72) [inline DataRepo/data/tests/submission_v3/sequences.tsv:4]: C. The `Sequence Number` column changed to `Sequence Name`: a comma-delimited combination of the composite unique values from `MSRunSequence`.  The values in this column will be used to populate dropdowns in the `Peak Annotation Details` sheet.

When the `Peak Annotation Details` sheet is used during the mzXML load, the values from this column will be read in, parsed, and used to query for the `MSRunSequence` record, in the same way other loaders retrieve related records, like the compound column in the tracers loader and the Tracer Name column in the infusates loader.

Note, the date-format of the date included in the concatenation will likely have to be controlled to work in the query, but for now, I'm not worrying about it.  That code does not yet exist.

File: DataRepo/loaders/sequences_loader.py [hunk 3]
```diff
@@ -1,26 +1,23 @@
...
 from DataRepo.models import LCMethod, MSRunSequence
-from DataRepo.utils.exceptions import RequiredColumnValueWhenNovel
+from DataRepo.utils.exceptions import InfileError, RecordDoesNotExist
 from DataRepo.utils.file_utils import string_to_datetime
 
 
 class SequencesLoader(TableLoader):
-    # TODO: 1. Implement a sequence accession composed of study code and sequence number
     # Header keys (for convenience use only).  Note, they cannot be used in the namedtuple() call.  Literal required.
-    ID_KEY = "ID"  # See TODO 1 above
```
Comment (hepcat72) [inline DataRepo/loaders/sequences_loader.py:17]: C. With the new `Sequence Name` column, the idea of using a prepended "study code" is obsolete.

File: DataRepo/data/tests/submission_v3/sequences.tsv (hunk 2 above)
Comment (hepcat72) [reply DataRepo/data/tests/submission_v3/sequences.tsv:4]: The idea to use a delimited string was inspired by other loaders that do with sort of related record retrieval.  I realized that those other loaders were using a unique key, and with the MSRunSequence table, that is just a "composite key".

Also note that the order of the values is important.  The value will be constructed using an excel formula and it will be a "read only" column.

File: DataRepo/loaders/sequences_loader.py [hunk 4]
```diff
@@ -1,26 +1,23 @@
 from collections import namedtuple
```
Comment (hepcat72) [inline DataRepo/loaders/sequences_loader.py:1]: C. Changes in this file, put simply, are:
- Remove the LC Protocol related columns/code
- Change the `Sequence Number` column to `Sequence Name` (a comma-delimited key combo to be able to query for `MSRunSequence` records), populated by excel formula

File: DataRepo/loaders/table_loader.py [hunk 5]
```diff
@@ -19,6 +19,7 @@
     DuplicateValues,
```
Comment (hepcat72) [inline DataRepo/loaders/table_loader.py:19]: C. While doing this work, I noted that since I had neglected to update the `Defaults` sheet, which had defaults for the LC protocol related columns in the sequences sheet, there were multiple errors that were a little confusing.  There only needed to be 1 consolidated error, so I changed the buffering of the key errors I was seeing to raise a combined `KeyError` and catch it to repackage as a more contextualized error.  I also updated a doc-string to explain that defaults are only returned if they have a non-empty value, which I'd also realized was confusing.

File: DataRepo/tests/utils/test_exceptions.py [hunk 6]
```diff
@@ -1003,3 +1005,13 @@ def test_MissingDataAdded(self):
...
+
+    def test_RecordDoesNotExist(self):
+        rdne = RecordDoesNotExist(
+            model=get_model_by_name("Tissue"),
+            query_dict={"name": "invalid"},
+        )
+        self.assertEqual(
+            "Tissue record matching {'name': 'invalid'} from the load file data does not exist.",
+            str(rdne),
+        )
```
Comment (hepcat72) [inline DataRepo/tests/utils/test_exceptions.py:1017]: C. I needed to provide file context for the `LCMethod.DoesNotExist` exception, and I realized that I could make a single exception for any model record.

File: DataRepo/utils/exceptions.py [hunk 7]
```diff
@@ -1460,23 +1460,30 @@ def __init__(
         differences,
         rec_dict=None,
         message=None,
+        derived=False,
```
Comment (hepcat72) [inline DataRepo/utils/exceptions.py:1463]: C. I wanted to be able to create this exception without having tried to load the database.  This exception class was originally created to handle/repackage `IntegrityErrors`, but in this instance, I was creating a conflicting value error relating to an automatically generated field value (`LCMethod.name`), and I didn't want the exception to reference a DB load attempt, as if the record existed, which it didn't necessarily, so I added the `derived` argument to be able to provide the correct context.

File: DataRepo/utils/exceptions.py [hunk 8]
```diff
@@ -1460,23 +1460,30 @@ def __init__(
...
         **kwargs,
     ):
         """Constructor
 
         Args:
```
Comment (hepcat72) [inline DataRepo/utils/exceptions.py:1468]: C. Improved the docstring.

File: DataRepo/tests/loaders/test_lcprotocols_loader.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        self.assertEqual(1, LCMethod.objects.count())
+        lcr = LCMethod.objects.first()
+        self.assertEqual("polar-HILIC-25-min", lcr.name)
+
+    def test_get_or_create_lc_method(self):
+        for _, row in self.TEST_DF.iterrows():
+            break
+        ll = LCProtocolsLoader()
+        rec, created = ll.get_or_create_lc_method(row)
+        self.assertEqual(0, len(ll.aggregated_errors_object.exceptions))
```
Comment (lparsons) [inline DataRepo/tests/loaders/test_lcprotocols_loader.py:40]: Comment: Handling exceptions outside of the `get_or_create` would eliminate the need for this check (which is more intuitive to me).

**Review by lparsons — APPROVED** [review]
The main change in this PR is to separate LC Protocols into a separate sheet and separate loading code instead of parsing them from supplied values on the Sequences sheet.

In addition, the PR also changes Sequence Number to Sequence Name and uses that on the Peak Annotation Details sheet which makes for a more intuitive drop-down. 

Finally, `KeyErrors` are consolidated to simplify error reporting.

I had a few comments and questions regarding some choices in the code, but no blocking issues. I also added on minor suggestion about exception handling.

File: DataRepo/loaders/lcprotocols_loader.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        try:
+            name = self.get_row_val(row, self.headers.NAME)
+            type = self.get_row_val(row, self.headers.TYPE)
+            raw_run_length = self.get_row_val(row, self.headers.RUNLEN)
+            description = self.get_row_val(row, self.headers.DESC)
+
+            # In case run_length was None, let's prevent an exception at the timedelta
+            if self.is_skip_row():
+                self.errored(LCMethod.__name__)
+                return rec, created
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/loaders/lcprotocols_loader.py:153]: That's a good question.  Perhaps you can help me improve the documentation.  Let me explain.

There is an instance attribute (a list) named `skip_row_indexes`.  Its intent is to facilitate the continued processing of input files amidst errors by skipping rows that we know will cause a cascade of errors.  The idea is to catch and report the root error and then proceed to find as many more unrelated/unique errors as we can find without bombarding the user with errors that would all go away if they address the root error - i.e. the one that started the cascade of errors.

There are various preemptive checks in both the base class (`TableLoader`) and in various derived classes that prevent that error-cascade by adding problematic row indexes to the list of `skip_row_indexes`.

The easiest example is the `get_row_val()` method.  It buffers an error and adds a row index to the `skip_row_indexes` list if the value in that column is `required` but `None`.  So if there's a required value missing, it buffers a `RequiredValue` error and adds a skip row index to avoid any number of errors down the line.

A derived class case is usually when there are multiple models being loaded from a single row or when any sort of preprocessing check occurs.  C.I.P. the compound synonym model.  The whole input file is preprocessed to check for synonyms in common with the primary compound names.  The compounds loader uses a call to `self.add_skip_row_index(index_list=syn_dict[syn])` to populate the `skip_row_indexes` when there is a problem.  Then in the actual loading loop, any rows marked to be skipped are skipped by checking `is_skip_row()`.

File: DataRepo/loaders/lcprotocols_loader.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+            # run_length is a required value (see DataRequiredValues), and is typed to be an int (see DataColumnTypes)
+            run_length = timedelta(minutes=raw_run_length)
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/loaders/lcprotocols_loader.py:156]: This is related to another pre-processing check, and I can see that the comment is insufficient.  The `TableLoader` class has a class attribute that describes the type of each column (`DataColumnTypes`).  In that attribute (a `dict`), `run_length` is described to be an `int`.  The idea here was that if there was an invalid type, an exception was buffered and the row was added to the `skip_row_indexes`...

However, this check is a bit outdated.  The type check used to be performed inside `get_row_val`, but some time back, I moved the type check out to the command class by querying for the `DataColumnTypes` and passing them to the `dtype` argument of the various pandas `read_*` methods.

So this effectively cannot happen anymore.  If there was a type error, it would have stopped execution before `load_data` was even called.  I think that I did that because if there was ever a type issue, it was the entire column that was wrong and nothing could be processed anyway.

In other words, the `timedelta` comment is obsolete here.  The only reason for the call to `is_skip_row` above is for missing required values.  We are not avoiding type errors here anymore.  That would have happened before execution got here.

[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/loaders/lcprotocols_loader.py:156]: - [ ] Remove the `DataColumnTypes` comment related to `timedelta`.

File: DataRepo/tests/loaders/test_lcprotocols_loader.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            df=df, file="DataRepo/data/tests/submission_v3/study.xlsx"
+        )
+        ll.load_data()
+        self.assertEqual(1, LCMethod.objects.count())
+        lcr = LCMethod.objects.first()
+        self.assertEqual("polar-HILIC-25-min", lcr.name)
+
+    def test_get_or_create_lc_method(self):
+        for _, row in self.TEST_DF.iterrows():
+            break
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/loaders/test_lcprotocols_loader.py:37]: Yes. I figured there was likely a way to do it, but I hadn't taken the time to look it up, so I'm glad you chimed in here.

- [ ] TODO: Use `self.TEST_DF.head(1)` instead of a lopp that just calls `break`.

File: DataRepo/tests/loaders/test_lcprotocols_loader.py (hunk 9 above)
Comment (hepcat72) [reply DataRepo/tests/loaders/test_lcprotocols_loader.py:40]: I agree.  I've already been thinking about modifying this design pattern.  In fact, I'd added a TODO comment in the `msrun_loader` PR for precisely what you suggest, because it became an awkward issue once I realized I needed an instance member of one loader inside another that called one of these methods like this, and it needed to merge its `AggregatedErrors` object.  So I already plan to make that change and will resolve this particular issue.

File: DataRepo/tests/loaders/test_lcprotocols_loader.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/loaders/test_lcprotocols_loader.py:37]: Huh.  Interestingly, neither of those suggestions work.  `.iloc[0]` applied an incompatible type (`numpy.int64`) and `.head(1)` doesn't set the row index (as `row.name`) used in the ensuing code.  It needs to have been generated by an iterator.

So I found [this solution](https://stackoverflow.com/a/23151722/2057516) (using the python3 version in the second comment), which is a bit better than what I *had*:

```
_, row = next(self.TEST_DF.iterrows())
```
